### PR TITLE
fix(iot-dev): Fix connection status reporting CONNECTED before all SDK states are CONNECTED

### PR DESCRIPTION
**Pull request by timtay-microsoft**
We were executing user facing callbacks with CONNECTED before we were executing SDK facing callbacks with the same state. As a result, users would try to do something like send telemetry as soon as CONNECTED state callback executes only to see an error saying that the deviceIO layer wasn't CONNECTED yet, too.

The solution is to make sure all SDK facing state callbacks are executed before user facing state callbacks are executed

#1348

### CODE REVIEW

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java [hunk 1]
```diff
@@ -1539,8 +1541,6 @@ private void updateStatus(IotHubConnectionStatus newConnectionStatus, IotHubConn
             {
                 this.multiplexingStateCallback.execute(newConnectionStatus, reason, throwable, this.multiplexingStateCallbackContext);
             }
-
-            this.deviceIOConnectionStatusChangeCallback.execute(newConnectionStatus, reason, throwable, null);
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:1543]: The user facing state callbacks were above this SDK facing state callback previously. Now this SDK facing state callback executes first

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java [hunk 2]
```diff
@@ -636,7 +636,15 @@ public void onDeviceSessionOpened(String deviceId)
...
-            this.listener.onMultiplexedDeviceSessionEstablished(this.connectionId, deviceId);
+
+            if (isMultiplexing)
+            {
+                // For non-multiplexing cases, the above logic for counting down the latch for the single device session
+                // will notify the open() thread that the reactor thread has successfully opened the device session.
+                // That will then execute the onConnectionEstablished callback on this same listener. Because of that,
+                // there is no listener callback to execute here for non-multiplexing cases.
+                this.listener.onMultiplexedDeviceSessionEstablished(this.connectionId, deviceId);
+            }
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java:647]: I found this bug while investigating this issue. We shouldn't have been calling "on**Multiplexed**DeviceSessionEstablished" when not multiplexing. It resulted in the SDK making multiple internal connection state updates when only one should have happened.

Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java:648]: This bug shouldn't have any meaningful impact on a user, but it did make debugging the connection state reporting confusing

**Review by abhipsaMisra — APPROVED** [review]

**Pull request merged**